presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+            }
+        }
+    }
+
+    /**
+     * This method can only be called when the table is known to exist
+     */
+    private TableSource getTableSource(String databaseName, String tableName)
+    {
+        checkReadable();
+        Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
+        if (tableAction == null) {
+            return TableSource.PRE_EXISTING_TABLE;
+        }
+        switch (tableAction.getType()) {
+            case ADD:
+                return TableSource.CREATED_IN_THIS_TRANSACTION;
+            case ALTER:
+                throw new IllegalStateException("Tables are never altered in the current implementation");
+            case DROP:
+                throw new TableNotFoundException(new SchemaTableName(databaseName, tableName));
+            case INSERT_EXISTING:
+                return TableSource.PRE_EXISTING_TABLE;
+            default:
+                throw new IllegalStateException("Unknown action type");
+        }
+    }
+
+    public synchronized Optional<List<String>> getAllViews(String databaseName)
+    {
+        checkReadable();
+        if (!tableActions.isEmpty()) {
+            throw new UnsupportedOperationException("Listing all tables after adding/dropping/altering tables/views in a transaction is not supported");
+        }
+        return delegate.getAllViews(databaseName);
+    }
+
+    /**
+     * {@code currentLocation} needs to be supplied if a writePath exists for the table.
+     */
+    public synchronized void createTable(ConnectorSession session, Table table, PrincipalPrivilegeSet principalPrivilegeSet, Optional<Path> currentPath)
+    {
+        setShared();
+        checkNoPartitionAction(table.getDatabaseName(), table.getTableName());
+        SchemaTableName schemaTableName = new SchemaTableName(table.getDatabaseName(), table.getTableName());
+        Action<TableAndMore> oldTableAction = tableActions.get(schemaTableName);
+        TableAndMore tableAndMore = new TableAndMore(table, Optional.of(principalPrivilegeSet), currentPath, Optional.empty());
+        if (oldTableAction == null) {
+            tableActions.put(schemaTableName, new Action<>(ActionType.ADD, tableAndMore, session.getUser(), session.getQueryId()));
+            return;
+        }
+        switch (oldTableAction.getType()) {
+            case DROP:
+                throw new PrestoException(TRANSACTION_CONFLICT, "Dropping and then recreating the same table in a transaction is not supported");
+            case ADD:
+            case ALTER:
+            case INSERT_EXISTING:
+                throw new TableAlreadyExistsException(schemaTableName);
+            default:
+                throw new IllegalStateException("Unknown action type");
+        }
+    }
+
+    public synchronized void dropTable(ConnectorSession session, String databaseName, String tableName)
+    {
+        setShared();
+        checkNoPartitionAction(databaseName, tableName);
```

Comment:
I'm not sure why you have this restriction here.  If we are dropping a table, can't we simply remove all of the partition actions?